// action-sheet-rn/lib/as.android.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var react_1 = __importStar(require("react"));
var react_native_1 = require("react-native");
var helpers_1 = require("./helpers");
var ActionSheetAndroid = /** @class */ (function (_super) {
    __extends(ActionSheetAndroid, _super);
    function ActionSheetAndroid() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { deviceHeight: 0, childrenHeight: 0 };
        _this.positionY = new react_native_1.Animated.Value(0);
        _this.opacity = new react_native_1.Animated.Value(0);
        _this.open = function () {
            return requestAnimationFrame(function () {
                react_native_1.Animated.parallel([
                    react_native_1.Animated.timing(_this.positionY, {
                        duration: 150,
                        toValue: 0,
                        useNativeDriver: true
                    }),
                    react_native_1.Animated.timing(_this.opacity, {
                        duration: 150,
                        toValue: 0.68,
                        useNativeDriver: true
                    })
                ]).start();
            });
        };
        _this.close = function (callback) {
            return requestAnimationFrame(function () {
                react_native_1.Animated.parallel([
                    react_native_1.Animated.timing(_this.positionY, {
                        duration: 150,
                        toValue: _this.state.childrenHeight,
                        useNativeDriver: true
                    }),
                    react_native_1.Animated.timing(_this.opacity, {
                        duration: 150,
                        toValue: 0,
                        useNativeDriver: true
                    })
                ]).start(function (e) {
                    if (e.finished) {
                        callback();
                    }
                });
            });
        };
        return _this;
    }
    ActionSheetAndroid.prototype.componentDidMount = function () {
        var childrenHeight = react_1["default"].Children.toArray(this.props.children).length * 56;
        this.setState(function () { return ({
            childrenHeight: childrenHeight,
            deviceHeight: react_native_1.Dimensions.get("window").height
        }); });
        this.positionY.setValue(childrenHeight);
        this.open();
    };
    ActionSheetAndroid.prototype.render = function () {
        var _this = this;
        var _a = this.props, children = _a.children, title = _a.title;
        var childrenArray = react_1["default"].Children.toArray(children);
        var translateY = {
            transform: [{ translateY: this.positionY }]
        };
        var cancelItem = childrenArray.find(helpers_1.isCancel);
        var transformedChildren = childrenArray
            .filter(helpers_1.isNotCancel)
            .map(function (child) {
            return react_1["default"].cloneElement(child, child.props.onPress && {
                onPress: function () { return _this.close(child.props.onPress); }
            });
        });
        return (react_1["default"].createElement(react_native_1.Modal, { transparent: true },
            react_1["default"].createElement(react_native_1.Animated.View, { style: [styles.container, { opacity: this.opacity }] },
                react_1["default"].createElement(react_native_1.TouchableWithoutFeedback, { onPress: function () { return cancelItem && _this.close(cancelItem.props.onPress); } },
                    react_1["default"].createElement(react_native_1.View, { style: styles.escape }))),
            react_1["default"].createElement(react_native_1.Animated.View, { style: [styles.card, translateY] },
                title && (react_1["default"].createElement(react_native_1.View, { style: styles.titleContainer },
                    react_1["default"].createElement(react_native_1.Text, { style: styles.title }, title))),
                transformedChildren)));
    };
    return ActionSheetAndroid;
}(react_1.Component));
exports.ActionSheetAndroid = ActionSheetAndroid;
var styles = react_native_1.StyleSheet.create({
    card: {
        backgroundColor: "white",
        bottom: 0,
        left: 0,
        position: "absolute",
        right: 0
    },
    container: {
        backgroundColor: "black",
        bottom: 0,
        left: 0,
        position: "absolute",
        right: 0,
        top: 0
    },
    escape: { flex: 1 },
    title: {
        color: "#0000008a",
        fontSize: 16
    },
    titleContainer: {
        height: 56,
        justifyContent: "center",
        paddingHorizontal: 16
    }
});
