// action-sheet-rn/lib/as.ios.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var react_1 = __importDefault(require("react"));
var react_native_1 = require("react-native");
var helpers_1 = require("./helpers");
var ActionSheetIos = /** @class */ (function (_super) {
    __extends(ActionSheetIos, _super);
    function ActionSheetIos() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ActionSheetIos.prototype.componentDidMount = function () {
        var _a = this.props, children = _a.children, message = _a.message, title = _a.title, color = _a.color;
        var childrenArray = react_1["default"].Children.toArray(children).filter(helpers_1.isPressable);
        var texts = childrenArray.map(function (child) { return child.props.children; });
        var cancelIndex = childrenArray.findIndex(helpers_1.isCancel);
        var removeIndex = childrenArray.findIndex(helpers_1.isRemove);
        react_native_1.ActionSheetIOS.showActionSheetWithOptions({
            cancelButtonIndex: cancelIndex,
            destructiveButtonIndex: removeIndex,
            message: message,
            options: texts,
            tintColor: color,
            title: title
        }, function (buttonIndex) {
            var child = childrenArray[buttonIndex];
            return child && child.props.onPress && child.props.onPress();
        });
    };
    ActionSheetIos.prototype.render = function () {
        return null;
    };
    return ActionSheetIos;
}(react_1["default"].Component));
exports.ActionSheetIos = ActionSheetIos;
