// action-sheet-rn/lib/sheetItem.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var react_1 = __importDefault(require("react"));
var react_native_1 = require("react-native");
exports.SheetItem = function (_a) {
    var children = _a.children, onPress = _a.onPress, Icon = _a.Icon;
    return (react_1["default"].createElement(react_native_1.TouchableNativeFeedback, { onPress: onPress, background: react_native_1.TouchableNativeFeedback.Ripple("#cecece") },
        react_1["default"].createElement(react_native_1.View, { style: styles.sheet },
            Icon && react_1["default"].createElement(react_native_1.View, { style: styles.icon }, Icon),
            react_1["default"].createElement(react_native_1.Text, { style: styles.text }, children))));
};
var styles = react_native_1.StyleSheet.create({
    icon: {
        marginRight: 32
    },
    sheet: {
        alignItems: "center",
        flexDirection: "row",
        height: 56,
        paddingHorizontal: 16
    },
    text: {
        color: "#000000de",
        flex: 1,
        fontSize: 16
    }
});
